src/mlpack/methods/gmm/gmm_diag_impl.hpp

```diff
@@ -0,0 +1,219 @@
+/**
+ * @author Kim SangYeon
+ * @file gmm_diag_impl.hpp
+ *
+ * Implementation of template-based DiagonalGMM methods.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_GMM_DIAG_GMM_IMPL_HPP
+#define MLPACK_METHODS_GMM_DIAG_GMM_IMPL_HPP
+
+// In case it hasn't already been included.
+#include "gmm_diag.hpp"
+
+namespace mlpack {
+namespace gmm {
+
+/**
+ * Fit the DiagonalGMM to the given observations.
+ */
+template<typename FittingType>
+double DiagonalGMM::Train(const arma::mat& observations,
+                          const size_t trials,
+                          const bool useExistingModel,
+                          FittingType fitter)
+{
+  // To use armadillo gmm_diag class, we create fitter for DiagonalConstraint.
+  EMFit<typename std::remove_reference<decltype(fitter.Clusterer())>::type,
+        DiagonalConstraint> armaFitter;
```

Comment:
If this is restricted to using Armadillo's fitter, I'd be fine with using a template parameter `InitialClusteringType` instead of `FittingType`; then you don't need to extract it from `FittingType`. :+1:

Reply:
As you said, `InitialClusteringType` would be better. Thanks. :)